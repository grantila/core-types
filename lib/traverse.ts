import type { NodePath, NodeType } from './types.js'

export interface TraverseCallbackArgument
{
	node: NodeType;
	rootNode: NodeType;
	path: NodePath;
	parentProperty?: string;
	parentNode?: NodeType;
	index?: string | number;
	required?: boolean;
}

export type TraverseCallback = ( arg: TraverseCallbackArgument ) => void;
export type SomeCallback = ( arg: TraverseCallbackArgument ) => boolean;

class StopError extends Error { }

export function traverse( node: NodeType, cb: TraverseCallback )
{
	function makeNewArg< T extends NodeType >(
		arg: TraverseCallbackArgument,
		parentNode: T,
		parentProperty: keyof T,
		index?: string | number,
		required?: boolean,
		newNode?: NodeType
	)
	: TraverseCallbackArgument
	{
		const node =
			newNode !== undefined
			? newNode
			: index === undefined
			? parentNode[ parentProperty ]
			: ( parentNode[ parentProperty ] as any)[ index ];

		const newPath: NodePath = [
			...arg.path,
			parentProperty as string,
			...( index === undefined ? [ ] : [ index ] ),
		];

		const newValues: Partial< TraverseCallbackArgument > = {
			node,
			path: newPath,
			parentNode,
			parentProperty: parentProperty as string,
			index,
			required,
		};

		return Object.assign( { }, arg, newValues );
	}

	function recurse( arg: TraverseCallbackArgument, cb: TraverseCallback )
	{
		cb( arg );

		const { node } = arg;

		if ( node.type === 'array' )
			recurse(
				makeNewArg< typeof node >( arg, node, 'elementType' ),
				cb
			);
		else if ( node.type === 'tuple' )
		{
			node.elementTypes.forEach( ( _, i ) =>
				recurse(
					makeNewArg< typeof node >( arg, node, 'elementTypes', i ),
					cb
				)
			);
			if ( typeof node.additionalItems === 'object' )
				recurse(
					makeNewArg< typeof node >( arg, node, 'additionalItems' ),
					cb
				);
		}
		else if ( node.type === 'object' )
		{
			for ( const prop of Object.keys( node.properties ) )
				recurse(
					makeNewArg< typeof node >(
						arg,
						node,
						'properties',
						prop,
						node.properties[ prop ].required,
						node.properties[ prop ].node
					),
					cb
				);

			if ( typeof node.additionalProperties === 'object' )
				recurse(
					makeNewArg< typeof node >(
						arg,
						node,
						'additionalProperties'
					),
					cb
				);
		}
		else if ( node.type === 'and' )
			node.and.forEach( ( _, i ) =>
				recurse(
					makeNewArg< typeof node >( arg, node, 'and', i ),
					cb
				)
			);
		else if ( node.type === 'or' )
			node.or.forEach( ( _, i ) =>
				recurse(
					makeNewArg< typeof node >( arg, node, 'or', i ),
					cb
				)
			);
	}

	const arg: TraverseCallbackArgument = {
		node,
		rootNode: node,
		path: [ ],
	};

	recurse( arg, cb );
}

export function some( node: NodeType, cb: SomeCallback )
{
	try
	{
		traverse( node, arg =>
		{
			if ( cb( arg ) )
				throw new StopError( );
		} );
	}
	catch ( err )
	{
		if ( err instanceof StopError )
			return true;
		throw err;
	}
	return false;
}
