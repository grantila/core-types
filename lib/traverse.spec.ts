import { type TraverseCallbackArgument, some, traverse } from './traverse.js'
import type { NodeType } from './types.js'


const root: NodeType = {
	type: 'object',
	title: 'root',
	additionalProperties: { type: 'string', title: 'o-ap' },
	properties: {
		any: { required: true, node: {
			title: 'o-any',
			type: 'any'
		}, },
		null: { required: true, node: {
			title: 'o-null',
			type: 'null'
		}, },
		string: { required: true, node: {
			title: 'o-string',
			type: 'string'
		}, },
		number: { required: true, node: {
			title: 'o-number',
			type: 'number'
		}, },
		integer: { required: false, node: {
			title: 'o-integer',
			type: 'integer'
		}, },
		boolean: { required: true, node: {
			title: 'o-boolean',
			type: 'boolean'
		}, },
		and: {
			required: true,
			node: {
				title: 'o-and',
				type: 'and',
				and: [ { type: 'number', title: 'o-and-number' } ],
			},
		},
		or: {
			required: true,
			node: {
				title: 'o-or',
				type: 'or',
				or: [ { type: 'number', title: 'o-or-number' } ],
			},
		},
		ref: {
			required: true,
			node: {
				title: 'o-ref',
				type: 'ref',
				ref: 'the-ref',
			},
		},
		object: {
			required: true,
			node: {
				title: 'o-object',
				type: 'object',
				properties: {
					foo: {
						required: false,
						node: {
							type: 'integer',
							title: 'o-o-foo-integer',
						},
					},
				},
				additionalProperties: { type: 'any', title: 'o-o-ap' },
			},
		},
		array: {
			required: false,
			node: {
				title: 'o-array',
				type: 'array',
				elementType: { type: 'null', title: 'o-array-null' },
			},
		},
		tuple: {
			required: true,
			node: {
				title: 'o-tuple',
				type: 'tuple',
				elementTypes: [
					{ type: 'number', title: 'o-tuple-0-number' },
					{ type: 'boolean', title: 'o-tuple-0-boolean' },
				],
				additionalItems: { type: 'any', title: 'o-tuple-any' },
				minItems: 1,
			},
		},
	},
};

describe( "traverse", ( ) =>
{
	type TestObject =
		{ title?: string }
		&
		Pick<
			TraverseCallbackArgument,
			| 'path'
			| 'required'
			| 'parentProperty'
			| 'index'
		>
		&
		Partial< Pick< TraverseCallbackArgument, 'rootNode' > >;

	const testObjectSorter = ( a: TestObject, b: TestObject ) =>
	{
		if ( a.path.length < b.path.length )
			return -1;
		else if ( a.path.length > b.path.length )
			return 1;
		return JSON.stringify( a ).localeCompare( JSON.stringify( b ) );
	};

	it( "should traverse deeply", ( ) =>
	{
		const visited: Array< TestObject > = [ ];
		traverse(
			root,
			( { node, path, required, parentProperty, index, rootNode } ) =>
			{
				const { title } = node;
				visited.push( {
					title,
					path,
					required,
					parentProperty,
					index,
					rootNode,
				} );
			}
		);

		for ( const elem of visited )
		{
			expect( elem.rootNode ).toBe( root );
			delete elem.rootNode;
		}

		visited.sort( testObjectSorter );

		expect( visited ).toMatchSnapshot( );
	} );

	it( "should traverse without additionals", ( ) =>
	{
		const visited: Array< TestObject > = [ ];
		const root: NodeType = {
			type: 'object',
			properties: {
				p: { required: false, node: {
					type: 'object',
					properties: {
						foo: { required: true, node: { type: 'boolean' } },
					},
					additionalProperties: false,
				} },
				tup: { required: true, node: {
					type: 'tuple',
					elementTypes: [ { type: 'null' } ],
					minItems: 0,
					additionalItems: false,
				}, },
			},
			additionalProperties: true,
		};
		traverse(
			root,
			( { node, path, required, parentProperty, index } ) =>
			{
				visited.push( {
					title: node === root ? '' : JSON.stringify( node ),
					path,
					required,
					parentProperty,
					index,
				} );
			}
		);

		visited.sort( testObjectSorter );

		expect( visited ).toMatchSnapshot( );
	} );
} );

describe( "some", ( ) =>
{
	it( "should not find what's not there", ( ) =>
	{
		const found = some(
			root,
			( { node } ) =>
				node.title === 'o-ref' && node.type === 'integer'
		);

		expect( found ).toBe( false );
	} );

	it( "should not find what's actually there", ( ) =>
	{
		const found = some(
			root,
			( { node } ) =>
				node.title === 'o-ref' && node.type === 'ref'
		);

		expect( found ).toBe( true );
	} );

	it( "should forward errors", ( ) =>
	{
		const find = ( ) => some(
			root,
			( { node } ) =>
			{
				if ( node.title === 'o-ref' && node.type === 'ref' )
					throw new Error( 'some error' );
				return false;
			}
		);

		expect( find ).toThrowError( 'some error' );
	} );
} );
