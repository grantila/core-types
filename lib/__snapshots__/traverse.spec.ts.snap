// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverse should traverse deeply 1`] = `
[
  {
    "index": undefined,
    "parentProperty": undefined,
    "path": [],
    "required": undefined,
    "title": "root",
  },
  {
    "index": undefined,
    "parentProperty": "additionalProperties",
    "path": [
      "additionalProperties",
    ],
    "required": undefined,
    "title": "o-ap",
  },
  {
    "index": "and",
    "parentProperty": "properties",
    "path": [
      "properties",
      "and",
    ],
    "required": true,
    "title": "o-and",
  },
  {
    "index": "any",
    "parentProperty": "properties",
    "path": [
      "properties",
      "any",
    ],
    "required": true,
    "title": "o-any",
  },
  {
    "index": "array",
    "parentProperty": "properties",
    "path": [
      "properties",
      "array",
    ],
    "required": false,
    "title": "o-array",
  },
  {
    "index": "boolean",
    "parentProperty": "properties",
    "path": [
      "properties",
      "boolean",
    ],
    "required": true,
    "title": "o-boolean",
  },
  {
    "index": "integer",
    "parentProperty": "properties",
    "path": [
      "properties",
      "integer",
    ],
    "required": false,
    "title": "o-integer",
  },
  {
    "index": "null",
    "parentProperty": "properties",
    "path": [
      "properties",
      "null",
    ],
    "required": true,
    "title": "o-null",
  },
  {
    "index": "number",
    "parentProperty": "properties",
    "path": [
      "properties",
      "number",
    ],
    "required": true,
    "title": "o-number",
  },
  {
    "index": "object",
    "parentProperty": "properties",
    "path": [
      "properties",
      "object",
    ],
    "required": true,
    "title": "o-object",
  },
  {
    "index": "or",
    "parentProperty": "properties",
    "path": [
      "properties",
      "or",
    ],
    "required": true,
    "title": "o-or",
  },
  {
    "index": "ref",
    "parentProperty": "properties",
    "path": [
      "properties",
      "ref",
    ],
    "required": true,
    "title": "o-ref",
  },
  {
    "index": "string",
    "parentProperty": "properties",
    "path": [
      "properties",
      "string",
    ],
    "required": true,
    "title": "o-string",
  },
  {
    "index": "tuple",
    "parentProperty": "properties",
    "path": [
      "properties",
      "tuple",
    ],
    "required": true,
    "title": "o-tuple",
  },
  {
    "index": undefined,
    "parentProperty": "elementType",
    "path": [
      "properties",
      "array",
      "elementType",
    ],
    "required": undefined,
    "title": "o-array-null",
  },
  {
    "index": undefined,
    "parentProperty": "additionalProperties",
    "path": [
      "properties",
      "object",
      "additionalProperties",
    ],
    "required": undefined,
    "title": "o-o-ap",
  },
  {
    "index": undefined,
    "parentProperty": "additionalItems",
    "path": [
      "properties",
      "tuple",
      "additionalItems",
    ],
    "required": undefined,
    "title": "o-tuple-any",
  },
  {
    "index": 0,
    "parentProperty": "and",
    "path": [
      "properties",
      "and",
      "and",
      0,
    ],
    "required": undefined,
    "title": "o-and-number",
  },
  {
    "index": "foo",
    "parentProperty": "properties",
    "path": [
      "properties",
      "object",
      "properties",
      "foo",
    ],
    "required": false,
    "title": "o-o-foo-integer",
  },
  {
    "index": 0,
    "parentProperty": "or",
    "path": [
      "properties",
      "or",
      "or",
      0,
    ],
    "required": undefined,
    "title": "o-or-number",
  },
  {
    "index": 1,
    "parentProperty": "elementTypes",
    "path": [
      "properties",
      "tuple",
      "elementTypes",
      1,
    ],
    "required": undefined,
    "title": "o-tuple-0-boolean",
  },
  {
    "index": 0,
    "parentProperty": "elementTypes",
    "path": [
      "properties",
      "tuple",
      "elementTypes",
      0,
    ],
    "required": undefined,
    "title": "o-tuple-0-number",
  },
]
`;

exports[`traverse should traverse without additionals 1`] = `
[
  {
    "index": undefined,
    "parentProperty": undefined,
    "path": [],
    "required": undefined,
    "title": "",
  },
  {
    "index": "p",
    "parentProperty": "properties",
    "path": [
      "properties",
      "p",
    ],
    "required": false,
    "title": "{"type":"object","properties":{"foo":{"required":true,"node":{"type":"boolean"}}},"additionalProperties":false}",
  },
  {
    "index": "tup",
    "parentProperty": "properties",
    "path": [
      "properties",
      "tup",
    ],
    "required": true,
    "title": "{"type":"tuple","elementTypes":[{"type":"null"}],"minItems":0,"additionalItems":false}",
  },
  {
    "index": "foo",
    "parentProperty": "properties",
    "path": [
      "properties",
      "p",
      "properties",
      "foo",
    ],
    "required": true,
    "title": "{"type":"boolean"}",
  },
  {
    "index": 0,
    "parentProperty": "elementTypes",
    "path": [
      "properties",
      "tup",
      "elementTypes",
      0,
    ],
    "required": undefined,
    "title": "{"type":"null"}",
  },
]
`;
